### PR TITLE
Add index object meta query

### LINKED ISSUE #3074: index object meta queries
Allow users to query the list of index objects using commit level meta queries.

### PR DESCRIPTION
**Pull request by mattnibs**
Add the meta query ":indices" that displays all indexed objects at
specific point in the commit history.

Closes #3074

### CODE REVIEW

File: lake/commits/actions.go [hunk 1]
```diff
@@ -91,3 +91,17 @@ func (a *AddIndex) String() string {
...
+	RuleID ksuid.KSUID `zng:"rule_id"`
+}
+
+func (d *DeleteIndex) String() string {
+	return fmt.Sprintf("DEL_INDEX %s/%s", d.RuleID, d.ID)
+}
+
+func (d *DeleteIndex) CommitID() ksuid.KSUID {
+	return d.Commit
+}
```
Comment (mattnibs) [inline lake/commits/actions.go:107]: The DeleteIndex stuff here isn't really used in this pr, but I always already adding AddIndex operations to Snapshots so I just included it in this pr. #3000 (coming soon!) will fully use this action.

File: lake/index/object.go [hunk 2]
```diff
@@ -8,17 +8,61 @@ import (
...
 }
 
 func (o Object) Path(path *storage.URI) *storage.URI {
 	return path.AppendPath(o.Rule.RuleID().String())
 	//	return xObjectPath(path, o.Rule, o.ID)
 }
+
+type Map map[ksuid.KSUID]ObjectRules
+
+func NewMap() Map { return make(Map) }
```
Comment (nwt) [inline lake/index/object.go:32]: `make(Map)` is only one character longer than `NewMap()`, so maybe drop this?

File: lake/index/object.go [hunk 3]
```diff
@@ -8,17 +8,61 @@ import (
...
+func (m Map) Insert(object *Object) {
+	rules, ok := m[object.ID]
+	if !ok {
+		rules = make(map[ksuid.KSUID]*Object)
+		m[object.ID] = rules
+	}
+	rules[object.Rule.RuleID()] = object
+}
+
+func (m Map) Remove(ruleID, id ksuid.KSUID) {
```
Comment (nwt) [inline lake/index/object.go:52]: `Delete`?

File: lake/ztests/index/apply-by-ip.yaml [hunk 4]
```diff
@@ -6,20 +6,32 @@ script: |
...
       {x:127.0.0.2}
+  - name: query.zed
+    data: |
+      from (
+        :index_rules => sort id;
+        test@main:indices => sort rule.id | cut o:=this;
+      )
+      | left join on id = o.rule.id
+      | count() by name,type
+      | sort name
```
Comment (nwt) [inline lake/ztests/index/apply-by-ip.yaml:24]: This works without the cut or the sort.
```suggestion
      from (
        :index_rules => sort id;
        test@main:indices => sort rule.id;
      )
      | left join on id = rule.id
      | count() by name,type
```

**Review by nwt — APPROVED** [review]

File: lake/commits/patch.go [hunk 5]
```diff
@@ -48,6 +49,11 @@ func (p *Patch) SelectAll() DataObjects {
 	return objects
 }
 
+func (p *Patch) SelectIndices(span extent.Span, o order.Which) []*index.Object {
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply lake/commits/patch.go:52]: Sure

File: lake/scheduler.go [hunk 6]
```diff
@@ -166,3 +167,14 @@ func ScanPartitions(ctx context.Context, snap commits.View, span extent.Span, o
 	}
 	return nil
 }
+
+func ScanIndices(ctx context.Context, snap commits.View, span extent.Span, o order.Which, ch chan<- index.Object) error {
+	for _, idx := range snap.SelectIndices(span, o) {
+		select {
+		case ch <- *idx:
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply lake/scheduler.go:174]: Just following an above Scan* method. Will fix.

File: lake/ztests/index/apply-by-ip.yaml (hunk 4 above)
Comment (mattnibs) [reply lake/ztests/index/apply-by-ip.yaml:24]: I had this but it was passing intermittently for me. It wasn't doing this for you?

Comment (nwt) [reply lake/ztests/index/apply-by-ip.yaml:24]: I have `while make TEST=TestZed/lake/ztests/index/apply-by-ip; do :; done` running with it and haven't seen any failures.

Comment (mattnibs) [reply lake/ztests/index/apply-by-ip.yaml]: Oh snap, duh this only returns once record.

**Pull request merged**